frontend/src/diagram/palette/ContextualMenu.tsx

```diff
@@ -10,47 +10,145 @@
-  imageURL: PropTypes.string,
-});
-const propTypes = {
-  tools: PropTypes.arrayOf(toolType).isRequired,
-  invokeTool: PropTypes.func.isRequired,
-  invokeClose: PropTypes.func.isRequired,
-};
+export const getConnectorToolsQuery = gql`
+  query getConnectorTools(
+    $editingContextId: ID!
+    $representationId: ID!
+    $sourceDiagramElementId: ID!
+    $targetDiagramElementId: ID!
+  ) {
+    viewer {
+      editingContext(editingContextId: $editingContextId) {
+        representation(representationId: $representationId) {
+          description {
+            ... on DiagramDescription {
+              connectorTools(
+                sourceDiagramElementId: $sourceDiagramElementId
+                targetDiagramElementId: $targetDiagramElementId
+              ) {
+                id
+                label
+                imageURL
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+`;
 
 /**
  * The component used to display a Contextual Menu.
  *
  * @hmarchadour
  */
-export const ContextualMenu = ({ tools, invokeTool, invokeClose }) => {
+export const ContextualMenu = ({
+  editingContextId,
+  representationId,
+  sourceDiagramElement,
+  targetDiagramElement,
+  invokeTool,
+  invokeClose,
+}: ContextualMenuProps) => {
+  const [{ value, context }, dispatch] = useMachine<ContextualMenuContext, ContextualMenuEvent>(contextualMenuMachine);
+
+  const { toast, contextualMenu } = value as SchemaValue;
+  const { connectorTools, message } = context;
+
+  const {
+    loading: connectorToolsLoading,
+    data: connectorToolsData,
+    error: connectorToolsError,
+  } = useQuery<GQLGetConnectorToolsData, GQLGetConnectorToolsVariables>(getConnectorToolsQuery, {
+    variables: {
+      editingContextId,
+      representationId,
+      sourceDiagramElementId: sourceDiagramElement.id,
+      targetDiagramElementId: targetDiagramElement.id,
+    },
+  });
+
+  useEffect(() => {
+    if (!connectorToolsLoading) {
+      const event: HandleConnectorToolsResultEvent = {
+        type: 'HANDLE_CONNECTOR_TOOLS_RESULT',
+        result: connectorToolsData,
+      };
+      dispatch(event);
+    }
+    if (connectorToolsError) {
+      const { message } = connectorToolsError;
+      const showToastEvent: ShowToastEvent = { type: 'SHOW_TOAST', message };
+      dispatch(showToastEvent);
+    }
+  });
+
   const onToolClick = useCallback(
     (tool) => {
       invokeTool(tool);
     },
     [invokeTool]
   );
+
+  const menuContent =
+    contextualMenu === 'loaded' ? (
+      <>
+        <div className={styles.menu} attrs-data-testid={`PopupMenu`}>
+          <div className={styles.menuEntries}>
+            {connectorTools.map((tool) => (
+              <Tool tool={tool} thumbnail={false} onClick={onToolClick} key={tool.id} />
+            ))}
+          </div>
+        </div>
+        <div className={toolSectionStyles.modal} onClick={() => invokeClose()}></div>
+      </>
+    ) : null;
```

Comment:
Wouldn't that make the menu appear event if there are no tools?

Reply:
It would, we could make the backend return an error if there is no tool otherwise.

Reply:
I don't think that we need an error but we could either not display the popup at all (is that the current behavior?) or even better display a simple a message indicating that there are no compatible tools